plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ElasticsearchQueryBuilder.java

```diff
@@ -77,62 +78,71 @@ public static QueryBuilder buildSearchQuery(TupleDomain<ElasticsearchColumnHandl
+            return queryBuilder;
         }
 
-        return buildTermQuery(boolQueryBuilder, columnName, domain, type);
+        return buildTermQuery(queryBuilder, columnName, domain, type);
     }
 
     private static QueryBuilder buildTermQuery(BoolQueryBuilder queryBuilder, String columnName, Domain domain, Type type)
     {
+        List<QueryBuilder> shouldList = new ArrayList<>();
         for (Range range : domain.getValues().getRanges().getOrderedRanges()) {
-            BoolQueryBuilder rangeQueryBuilder = new BoolQueryBuilder();
             Set<Object> valuesToInclude = new HashSet<>();
             checkState(!range.isAll(), "Invalid range for column: %s", columnName);
             if (range.isSingleValue()) {
                 valuesToInclude.add(range.getSingleValue());
+                shouldList.add(new TermQueryBuilder(columnName, getValue(type, range.getSingleValue())));
             }
             else {
+                RangeQueryBuilder rangeQueryBuilder = new RangeQueryBuilder(columnName);
                 if (!range.isLowUnbounded()) {
                     Object lowBound = getValue(type, range.getLowBoundedValue());
                     if (range.isLowInclusive()) {
-                        rangeQueryBuilder.filter(new RangeQueryBuilder(columnName).gte(lowBound));
+                        rangeQueryBuilder.gte(lowBound);
                     }
                     else {
-                        rangeQueryBuilder.filter(new RangeQueryBuilder(columnName).gt(lowBound));
+                        rangeQueryBuilder.gt(lowBound);
                     }
                 }
                 if (!range.isHighUnbounded()) {
                     Object highBound = getValue(type, range.getHighBoundedValue());
                     if (range.isHighInclusive()) {
-                        rangeQueryBuilder.filter(new RangeQueryBuilder(columnName).lte(highBound));
+                        rangeQueryBuilder.lte(highBound);
                     }
                     else {
-                        rangeQueryBuilder.filter(new RangeQueryBuilder(columnName).lt(highBound));
+                        rangeQueryBuilder.lt(highBound);
                     }
                 }
+                shouldList.add(rangeQueryBuilder);
             }
-
-            if (valuesToInclude.size() == 1) {
-                rangeQueryBuilder.filter(new TermQueryBuilder(columnName, getValue(type, getOnlyElement(valuesToInclude))));
-            }
-            queryBuilder.should(rangeQueryBuilder);
         }
 
         if (domain.isNullAllowed()) {
-            queryBuilder.should(new BoolQueryBuilder().mustNot(new ExistsQueryBuilder(columnName)));
+            shouldList.add(new BoolQueryBuilder().mustNot(new ExistsQueryBuilder(columnName)));
+        }
+
+        // only one should Clause, we can add it to filter
+        if (shouldList.size() == 1) {
+            queryBuilder.filter(shouldList.get(0));
+        }
+        else if (shouldList.size() > 1) {
```

Comment:
no, it will not be empty, because we already check the domain.getValues() is not null.
Even if it is empty, we will do not add any to queryBuilder, it's ok.